File: ts-sdk/src/resources/sessions.ts
// Sessions resource: create (login), revoke (logout), refresh, list, and mint
// scoped tokens.

import type { Transport } from "../http/transport.js";
import type {
  CreateTokenParams,
  CreateTokenResult,
  LoginParams,
  RefreshSessionResult,
  Session,
} from "../types/index.js";
import {
  parseActiveSessions,
  parseCreateTokenResult,
  parseLoginToken,
  parseRefreshSessionResult,
} from "../utils/parse.js";
import {
  validateCreateTokenParams,
  validateLoginParams,
  validateSessionId,
} from "../utils/validation.js";

const SESSIONS_BASE = "/api/v1/sessions";
const ACTIVE_SESSIONS_PATH = `${SESSIONS_BASE}/active`;
const REFRESH_SESSION_PATH = `${SESSIONS_BASE}/refresh`;
const CREATE_TOKEN_PATH = `${SESSIONS_BASE}/tokens`;

/**
 * Authenticate and return the bearer token on success.
 *
 * @throws {@link ValidationError} when `username` or `password` is empty.
 * @throws {@link AuthenticationError} (401) on bad credentials.
 */
export async function login(transport: Transport, params: LoginParams): Promise<string> {
  validateLoginParams(params);
  const raw = await transport.request<unknown>("POST", SESSIONS_BASE, {
    body: { username: params.username, password: params.password },
  });
  return parseLoginToken(raw);
}

/**
 * Revoke the current session. After this call the bearer token is invalid.
 * Allowed for any session (login or minted) — it can only ever revoke itself.
 *
 * @throws {@link AuthenticationError} (401) when the token is already expired or revoked.
 */
export async function logout(transport: Transport, token: string): Promise<void> {
  await transport.request<void>("DELETE", SESSIONS_BASE, { token });
}

/**
 * Extend the current session's expiry by the server's standard TTL, without
 * changing the token value. Subject to an absolute age cap enforced server-side —
 * eventually a session must re-authenticate via {@link login} regardless of how
 * often it's refreshed.
 *
 * @throws {@link AuthenticationError} (401) when the session can no longer be
 * refreshed (expired, revoked, or past the absolute age cap) — log in again.
 */
export async function refreshSession(
  transport: Transport,
  token: string,
): Promise<RefreshSessionResult> {
  const raw = await transport.request<unknown>("POST", REFRESH_SESSION_PATH, { token });
  return parseRefreshSessionResult(raw);
}

/**
 * Mint a new scoped token on behalf of the caller. Restricted to login-origin
 * sessions — a minted token can never mint another one.
 *
 * @throws {@link ValidationError} when `scope` is not "read" or "write".
 * @throws {@link AuthenticationError} (403) when called from a minted (non-login) session.
 */
export async function createToken(
  transport: Transport,
  token: string,
  params: CreateTokenParams,
): Promise<CreateTokenResult> {
  validateCreateTokenParams(params);
  const raw = await transport.request<unknown>("POST", CREATE_TOKEN_PATH, {
    token,
    body: { scope: params.scope },
  });
  return parseCreateTokenResult(raw);
}

/**
 * List the authenticated user's active (non-expired, non-revoked) sessions.
 *
 * @throws {@link AuthenticationError} (401) when the token is invalid or expired.
 */
export async function getActiveSessions(transport: Transport, token: string): Promise<Session[]> {
  const raw = await transport.request<unknown>("GET", ACTIVE_SESSIONS_PATH, { token });
  return parseActiveSessions(raw);
}

/**
 * Revoke one of the authenticated user's active sessions by its session id
 * (as returned by {@link getActiveSessions}) — not necessarily the caller's
 * own session. Unlike {@link logout}, which always revokes the session
 * behind the current bearer token, this can end a session on another device.
 * Restricted to login-origin sessions — a minted token can never revoke a
 * session other than itself (use {@link logout} for that).
 *
 * @throws {@link ValidationError} when `sessionId` is empty.
 * @throws {@link AuthenticationError} (403) when called from a minted (non-login) session.
 * @throws {@link APIError} (404) when no active session matches `sessionId` for this user.
 */
export async function revokeSession(
  transport: Transport,
  token: string,
  sessionId: string,
): Promise<void> {
  validateSessionId(sessionId);
  await transport.request<void>("DELETE", ACTIVE_SESSIONS_PATH, {
    token,
    body: { session_id: sessionId },
  });
}

File: e2e/internal/client/sessions.go
package client

import (
	"context"
	"net/http"
)

type MintedToken struct {
	Token     string `json:"token"`
	Scope     string `json:"scope"`
	ExpiresAt int64  `json:"expires_at"`
}

type ActiveSession struct {
	CreatedAt int64   `json:"created_at"`
	ExpiresAt int64   `json:"expires_at"`
	SessionID string  `json:"session_id"` // one-way token hash — the revoke key, never an auth credential
	Origin    string  `json:"origin"`
	Scope     string  `json:"scope"`
	UserAgent *string `json:"user_agent,omitempty"`
	IPAddress *string `json:"ip_address,omitempty"`
}

type createTokenRequest struct {
	Scope string `json:"scope"` // "read" or "write"
}

type revokeSessionRequest struct {
	SessionID string `json:"session_id"`
}

type refreshResponse struct {
	ExpiresAt int64 `json:"expires_at"`
}

// MintToken issues a scoped token from a login session. loginToken must come from Login — a
// minted token cannot mint another (API returns 403).
func (c *Client) MintToken(ctx context.Context, loginToken, scope string) (*MintedToken, error) {
	var resp MintedToken
	if err := c.do(ctx, http.MethodPost, "/sessions/tokens", loginToken,
		createTokenRequest{Scope: scope}, &resp, http.StatusCreated); err != nil {
		return nil, err
	}
	return &resp, nil
}

// RefreshSession extends the session behind token and returns its new expiry.
func (c *Client) RefreshSession(ctx context.Context, token string) (int64, error) {
	var resp refreshResponse
	if err := c.do(ctx, http.MethodPost, "/sessions/refresh", token, nil, &resp, http.StatusOK); err != nil {
		return 0, err
	}
	return resp.ExpiresAt, nil
}

func (c *Client) ListSessions(ctx context.Context, token string) ([]ActiveSession, error) {
	var out []ActiveSession
	if err := c.do(ctx, http.MethodGet, "/sessions/active", token, nil, &out, http.StatusOK); err != nil {
		return nil, err
	}
	return out, nil
}

// RevokeSession revokes the session identified by sessionID (an ActiveSession.SessionID).
// token must be a login session (403 otherwise). A missing session returns nil, not an error.
func (c *Client) RevokeSession(ctx context.Context, token, sessionID string) error {
	return c.do(ctx, http.MethodDelete, "/sessions/active", token,
		revokeSessionRequest{SessionID: sessionID}, nil, http.StatusOK, http.StatusNotFound)
}

// Logout revokes the session behind token.
func (c *Client) Logout(ctx context.Context, token string) error {
	return c.do(ctx, http.MethodDelete, "/sessions", token, nil, nil, http.StatusNoContent)
}
